internal/service/kubernetes/node_group.go

```diff
@@ -0,0 +1,296 @@
+package kubernetes
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/UpCloudLtd/terraform-provider-upcloud/internal/utils"
+	"github.com/UpCloudLtd/upcloud-go-api/v5/upcloud"
+	"github.com/UpCloudLtd/upcloud-go-api/v5/upcloud/request"
+	"github.com/UpCloudLtd/upcloud-go-api/v5/upcloud/service"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceNodeGroup() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Kubernetes node group. NOTE: this is an experimental feature in an alpha phase, the resource definition might change in the future.",
+		CreateContext: resourceNodeGroupCreate,
+		ReadContext:   resourceNodeGroupRead,
+		DeleteContext: resourceNodeGroupDelete,
+		UpdateContext: resourceNodeGroupUpdate,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"cluster": {
+				Description: idDescription,
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"node_count": {
+				Description:      "Amount of nodes to provision in the node group.",
+				Type:             schema.TypeInt,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.IntBetween(0, 16)),
+				Required:         true,
+			},
+			"name": {
+				Description:      "The name of the node group. Needs to be unique within a cluster.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validateResourceName,
+				ForceNew:         true,
+			},
+			"plan": {
+				Description: "The pricing plan used for the node group. Valid values available via `upcloud_kubernetes_plan` datasource field `description`.",
```

Comment:
Now that we do not have separate `K8S-` plans, this could mention that any server plan can be used and those can be listed with `upctl server plans`. This could be updated in later PR as well.

Reply:
PR [#340](https://github.com/UpCloudLtd/terraform-provider-upcloud/pull/340)